test/test_query.cpp

```diff
@@ -5308,4 +5309,91 @@ TEST(Query_AllocatorBug)
+    auto col_type = table->add_column(type_String, "type");
+    auto col_tags = table->add_column(type_String, "tags");
+    table->add_search_index(col_type);
+
+    for (int i = 0; i < 10000; i++) {
+        auto obj = table->create_object();
+        std::string type = types[i % 9];
+        std::string val = type + util::to_string(i % 10);
+        obj.set(col_type, val);
+        std::string tags;
+        for (int j = 0; j < 7; j++) {
+            tags += " TAG" + util::to_string(i % 500);
+        }
+        obj.set(col_tags, tags);
+    }
+
+    auto q1 = table->where().equal(col_type, "todo0", false);
+    auto t1 = steady_clock::now();
+    auto cnt = q1.count();
+    auto t2 = steady_clock::now();
+    CHECK_EQUAL(cnt, 112);
+    auto dur1 = duration_cast<microseconds>(t2 - t1).count();
+    // std::cout << "cnt: " << cnt << " dur1: " << dur1 << " us" << std::endl;
+
+    auto q2 = table->where().contains(col_tags, "tag0", false);
+    t1 = steady_clock::now();
+    cnt = q2.count();
+    t2 = steady_clock::now();
+    CHECK_EQUAL(cnt, 20);
+    auto dur2 = duration_cast<microseconds>(t2 - t1).count();
+    // std::cout << "cnt: " << cnt << " dur2: " << dur2 << " us" << std::endl;
+
+    Query q = q1.and_query(q2);
+    t1 = steady_clock::now();
+    cnt = q.count();
+    t2 = steady_clock::now();
+    CHECK_EQUAL(cnt, 3);
+    auto dur3 = duration_cast<microseconds>(t2 - t1).count();
+
+    // The duration of the combined query should be closer to the duration for
+    // the query using the index.
+    CHECK_GREATER(dur3, dur1);
+    CHECK_LESS(dur3, dur2 / 5);
+    // std::cout << "cnt: " << cnt << " dur3: " << dur3 << " us" << std::endl;
+}
+
+TEST(Query_SwitchNode)
+{
+    // In the first 500 objects col_foo have fewest matches and in the
+    // next 500 col_bar has the fewest. Repeated 5 times. The performance
+    // is slightly increased if the query engine alters between the two
+    // nodes as being the driving node.
+    Group g;
+    TableRef table = g.add_table("table");
+    auto col_foo = table->add_column(type_Float, "foo");
+    auto col_bar = table->add_column(type_Float, "bar");
+
+    for (int j = 0; j < 5; j++) {
+        for (int i = 0; i < 500; i++) {
+            auto obj = table->create_object();
+            obj.set(col_foo, float(i % 50));
+            obj.set(col_bar, float(i % 5));
+        }
+
+        for (int i = 0; i < 500; i++) {
+            auto obj = table->create_object();
+            obj.set(col_foo, float(i % 5));
+            obj.set(col_bar, float(i % 50));
+        }
+    }
+
+    auto q1 = table->where().equal(col_foo, 1.0f);
+    auto q2 = table->where().equal(col_bar, 1.0f);
+    Query q = q1.and_query(q2);
+    auto t1 = steady_clock::now();
+    auto cnt = q.count();
+    auto t2 = steady_clock::now();
+    CHECK_EQUAL(cnt, 100);
+    auto dur = duration_cast<microseconds>(t2 - t1).count();
+    std::cout << "cnt: " << cnt << " duration: " << dur << " us" << std::endl;
```

Comment:
can we infer a timing check here instead of printing diagnostics? Perhaps a crude check is: time(q1) + time(q2) > time(q)?

Reply:
I am not sure it would prove anything. I removed the printout.

Reply:
I agree it's difficult to construct a test for this type of change; not a blocker.